added vectorized functions

kswiftie opened the pull request:
Некоторые векторизованные функции работают только на ICC или Visual Studio C++. К ним относятся математические функции. Они используются в следующих функциях активации: exponential - exp
sigmoid - exp
softplus - log, exp
swish - exp
tanh - exp

kswiftie committed:
added vectorized functions

Некоторые векторизованные функции работают только на ICC или Visual Studio C++. К ним относятся математические функции. Они используются в следующих функциях активации:
exponential - exp
sigmoid - exp
softplus - log, exp
swish - exp
tanh - exp

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -265,7 +271,193 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+(_mm{funcname}_add_ps(_mm{funcname}_set1_ps(1.0f), _mm{funcname}_exp_ps(_mm{funcname}_mul_ps(_mm{funcname}_set1_ps(-1.0f), vsum)))));\n""",
+        "tanh": f"""__m{typename} vans = _mm{funcname}_div_ps(_mm{funcname}_sub_ps(_mm{funcname}_exp_ps(vsum), _mm{funcname}_exp_ps(_mm{funcname}_mul_ps(_mm{funcname}_set1_ps(-1.0f), vsum))), 
+_mm{funcname}_add_ps(_mm{funcname}_exp_ps(vsum), _mm{funcname}_exp_ps(_mm{funcname}_mul_ps(_mm{funcname}_set1_ps(-1.0f), vsum))));\n""",
+    } if vectorized_level != "avx512" \
```
Как будто в зависимости от уровня векторизации меняются не все активационные функции, стоит для уменьшения количества кода и наглядности поменять только различающиеся активационные функции.
К примеру создать `vectorized_d`, затем если необходимо заменить конкретные строки.
Также в текущем виде абсолютно нечитаемы функции, рекомендую их как-нибудь отформатировать, чтобы было проще читать

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -265,7 +271,193 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+
+def get_vectorized_names(vectorized_level: str) -> tuple:
+    intrisincs = ["sse", "avx", "avx512"]
+    vectorized_typename = {"none": "", "sse": "128", "avx": "256", "avx512": "512"}
```
Зачем здесь `"none"`? Насколько я понял, на вход может прийти только одно из значений массива intrisincs. В таком случае надо просто проверить, что если значение не содержится в нём, то произошла чрезвычайная ситуация и необходимо выбросить исключение

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -1,3 +1,4 @@
+import cpuinfo # needed to install to get info about available intrinsics information
```
Есть ли аналоги не требующие дополнительной установки python пакета? Если нет, то надо добавить соответствующий пакет в `requirements`

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -218,19 +219,24 @@ def feed_forward_step(
     bias_name: str,
     activation_func: str,
+    decorator_params: list = None,
+    vectorized_level: str = "none",
```
Зачем нужен `decorator_params`?
И не хватает документации (doc-string) для `vectorized_level`

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -218,19 +219,24 @@ def feed_forward_step(
+                    {right_name}[i] += {weight_name}[j][i] * {left_name}[j];
+                }}
+                {right_name}[i] += {bias_name}[i];
+                {activation_to_cpp_template(right_name + "[i]", activation_func, vectorized_level, decorator_params)}
```
У функции `activation_to_cpp_template` другая сигнатура

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -265,7 +271,193 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+    typename, funcname = get_vectorized_names(vectorized_level)
+    weights_for_st_setr = ""
+    weights_for_nd_setr = ""
+    step = f"{int(typename) // 32}"
```
Откуда волшебная константа 32? Было бы неплохо завести константу с таким значением и именем отражающим суть

Krekep commented on degann/networks/topology/tf_densenet.py:
```diff
@@ -305,13 +306,8 @@ def export_to_cpp(
-
-        \n"""
+        res = """#include <cmath>
+#include <vector>\n"""
```
А на что ты заменяешь `max(x, y)`?

Krekep reviewed (changes requested):
Глобально хорошо, но надо ещё отполировать

Krekep commented:
Информацию о том, какие компиляторы (не)поддерживают те или иные функции стоит внести в doc-string, к примеру функции `export_to_cpp`.
Касательно github actions --- необходимо, чтобы все проверки проходили.
Также в идеале реализовать тесты, проверяющие корректность экспорта в c++

kswiftie committed:
sync

kswiftie committed:
update

added test for vectorization, added a description of vectorization, and some corrections have been made to improve the readability of the code

kswiftie committed:
fix by workflows

kswiftie committed:
fixes and documentation

added fixes and documentation for functions connected with vectorization
also was fixed mistakes in C++ code like using std functions without std::

kswiftie committed:
updates

updated workflows, added example, fix code

kswiftie committed:
update tests and workflows

add g++ to tests.yaml
fix test_vectorization for work with github tests

kswiftie committed:
fix code style

kswiftie committed:
update compare_time.py

add comments

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -244,31 +474,354 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+
+def generate_vectorized_function(vectorized_level: str, activation_func: str) -> str:
+    """
+    This function creates vectorized code based on feed_forward_step at the vectorization level!= "none"
```
Лучше использовать не `!=`, а `is not "none"` или чтото похожее

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -244,31 +474,354 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+            abs_vsum = _mm{funcname}_add_ps(abs_vsum, _mm{funcname}_and_ps(vflag, _mm{funcname}_mul_ps(_mm{funcname}_set1_ps(-1.0f), vsum)));
+            abs_vsum = _mm{funcname}_add_ps(abs_vsum, _mm{funcname}_andnot_ps(vflag, vsum));
+            __m{typename} vans = _mm{funcname}_div_ps(vsum, _mm{funcname}_add_ps(abs_vsum, _mm{funcname}_set1_ps(1.0f)));\n""",
+        # "softmax": lambda x: ,
```
Такой активационной функции нет в библиотеке

Krekep commented on degann/networks/imodel.py:
```diff
@@ -277,7 +279,12 @@ def clear_history(self):
+        path: str,
+        array_type: str = "[]",
+        path_to_compiler: str = None,
+        vectorized_level: str = "none",
```
Не хватает документации для `vectorized_level`. Также неплохо было бы перечислить где-нибудь (можно здесь), какие уровни векторизации вообще поддерживаются

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -244,31 +474,354 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+    Parameters
+    ----------
+    vectorized_level: str
+        level of code vectorization
```
Здесь и ещё в нескольких местах с большой буквы стоит написать

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -244,31 +474,354 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+        + """}
+                    }\n\n"""
+    )
+    return res if vectorized_level != "none" else ""
```
Дублирование проверки, в начале функции уже есть проверка

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -244,31 +474,354 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
+def get_vectorized_names(vectorized_level: str) -> tuple:
+    """
+    This function returns names for vectorized functions based on the vectorization level
+    Parameters
```
Между описанием функции и описанием её аргументов/возвращаемых значений надо добавить пустую строку

Krekep commented on degann/networks/cpp_utils.py:
```diff
@@ -244,31 +474,354 @@ def activation_to_cpp_template(name: str, activation_name: str) -> str:
-    for key, value in values:
-        act_func = act_func.replace(key, str(value))
-    return act_func
+def create_main_func(type: str = "default", path: str = "") -> str:
```
Этого не должно быть  в библиотеке.
Если это необходимо для тестов, эту функцию можно добавить в папку tests

Библиотека должна позволять по нейронной сети строить функцию, которую уже можно встраивать куда захочет пользователь. Она не должна предоставлять полноценную программу. Если пользователю необходимо, то он напишет функцию `main` сам

Krekep commented on degann/networks/topology/tf_densenet.py:
```diff
@@ -400,13 +423,32 @@ def export_to_cpp(
             output_size, f"layer_{len(blocks) + 1}", "answer"
         )
 
+        if "layers" in kwargs and kwargs["layers"] != "":
```
Зачем это нужно?

Krekep commented on examples/compare_time.py:
Лучше переименовать этот файл, т.к. непонятно сравнение времени чего

Krekep commented on examples/readme.md:
```diff
@@ -0,0 +1,3 @@
+# Example of usages
+
+Main use cases --- [Google colab notebook](https://colab.research.google.com/drive/1xoFGedC0BiihVVohDnw3p4l5Xz0w4vLQ?usp=sharing)
```
Стоит добавить ссылку на гугл ноутбук в этот файл.

Krekep commented on examples/notebook_code.py:
С учётом того, что ноутбуков теперь два, этот файл стоит переименовать, чтобы было понятно, к чему он относится

Krekep commented on examples/example.ipynb:
Этот ноутбук стоит переименовать в `main_use_cases` видимо

Krekep reviewed (changes requested):
В основном замечания по мелочи, уже очень хорошо всё выглядит.
Из крупных замечаний --- функцию `main` надо убрать и нет кодогенерации для активационной функции `parabolic`. В [этой строке](https://github.com/Krekep/degann/blob/028ce98f5b49b4332515cc8a9b4073cb3fc45a7a/degann/networks/cpp_utils.py#L487C9-L487C123) для неё написана кодогенерация без векторизации.
Также стоит добавить ноутбук с примерами использования обновлённой функции `export_to_cpp` в папку `examples`.

kswiftie committed:
fixes

fixed documentation
moved some functions
slightly changed some functions

kswiftie committed:
fix tests

kswiftie committed:
updates

fixed some generate code
fixed one documentation
fixed link to google collab

kswiftie committed:
fix codestyle

kswiftie committed:
fixed documentation

Krekep committed:
Merge branch 'main' into main

kswiftie committed:
fixed activation funcs

kswiftie committed:
Merge branch 'main' of https://github.com/kswiftie/degann

kswiftie committed:
fixed

kswiftie committed:
fixes

fixed function in cpp_utils
deleted unnecessary files

Krekep reviewed (approved):
Хорошо.
